verilog/tools/ls/autoexpand.cc

```diff
@@ -204,32 +312,38 @@ const std::regex AutoExpander::autotemplate_type_re_{
 const std::regex AutoExpander::autotemplate_conn_re_{
     R"(\.\s*([^\s(]+?)\s*\(\s*([^\s(]+?)\s*\))"};
 
-const AutoExpander::Template *AutoExpander::Template::from(
-    const Map &templates, const absl::string_view type_id,
-    const int64_t instance_offset) {
-  const auto templates_it = templates.find(type_id);
-  const Template *matching_tmpl = nullptr;
-  if (templates_it != templates.end()) {
-    // Linear search for the matching template (there should be very few
-    // templates per type, often just one)
-    for (const auto &tmpl : templates_it->second) {
-      if (instance_offset < tmpl.offset) break;
-      matching_tmpl = &tmpl;
-    }
-  }
-  return matching_tmpl;
-}
-
-void AutoExpander::Module::EmitPortDeclarations(
+// AUTOINPUT/OUTPUT/INOUT/WIRE/REG regex breakdown:
+// (/\*\s* ... \s*\*/\s*?\n)                – starting comment
+// (?:\s*//.*\n)?                           – optional starting comment
+//                                            ("Beginning of automatic...")
+// (?: ... )?                               – an optional non-capturing group:
+//   ([\s\S]*?)?                            – any text (usually port
+//                                            declarations)
+//   (?:[^\S\r\n]*// End of automatics.*\n) – ended by an "End of automatics"
+//                                            comment
+constexpr const char *auto_declaration_re_fmt =
+    R"((/\*\s*%s\s*\*/\s*?\n)(?:\s*//.*\n)?(?:([\s\S]*?)?(?:[^\S\r\n]*// End of automatics.*\n))?)";
+const std::regex AutoExpander::autoinput_re_{
```

Comment:
mmh, global constants that require complicated set-up calls are often murky in the way things are sequenced and available (here, we assemble a global value ideally before main() with StrFormat(). And also create a complicated regex in addition to that). (the compiler/linker might not always be able to properly find a good sequence for these "Static Initialization Order Fiasco").

To solve I suggest to use the pattern where we have a function to call that then does the initialization (static variables in functions are only initialized on the first call):

```c++
const std::regex& autoinput_re() {
   static const std::regex re{absl::StrFormat(auto_declaration_re_fmt, "AUTOINOUT")};
  return re;
}
```

(also the other places where regex are statically initalized)

Reply:
I think this is problematic only for dependencies between translation units? Also, the `StrFormat` is part of the same expression as the regex construction. It depends on a string constant which needs no initialization.

(compile-time regex compilation would be nice)

Reply:
Unless `StrFormat()` is constexpr, I'd be wary to call it before everything is initialized.

It is a function provided by a different library that is not under our control (`absl::`), so we can't assume that it does not depend on some other global state that might not be ready yet.

You could make the string concatenation at safe without calling `StrFormat()`, something along the lines of:

```c++
#define MAKE_AUTO_REGEX(whatmatch) R"((/\*\s*)" whatmatch R"(\s*\*/\s*?\n)(?:\s*//.*\n)?(?:[\s\S]*?[^\S\r\n]*// End of automatics.*\n)?)"

const std::regex AutoExpander::autoinput_re_{MAKE_AUTO_REGEX("AUTOINPUT")};
//... more inits
#undef MAKE_AUTO_REGEX
```
One of the few cases where macros are actually kindof warranted...

Using global static std::regex is technically also not ok, because it is not `is_trivially_destructible`, which is outside of our [style guide suggestions](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables). Though I'd say it is benign and probably something we can ignore in this case for now.

So if you can avoid the StrFormat() call, I think it's ok.

(And yes, it would be good if `std::regex` was constexpr, but alas, it throws. So unless that is fixed in the standard, we won't see that... (though I have seen people made impressive `constexpr` regular expression matchers).)

Reply:
Seems like local statics are also disallowed by the guidelines. I think the only way to conform to them is to make `AutoExpander` a persistent object that is constructed by the language server entry point, and the regexes its fields.

I think I'll use the macro as it's more concise and consistent with the other regexes.